drivers/flash: STM32: A generic change and enabling flash support.

ikmdani opened the pull request:
This commit makes the flash test generic for both qspi-flash and
soc-flash. For STM32 based platforms, the flast test will look for
storage partition in the soc-flash for testing.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

nvlsianpu reviewed (commented):
Usage of the storage partition for all SoC embedded flash would be better solution.

nvlsianpu commented on tests/drivers/flash/testcase.yaml:
```diff
@@ -11,3 +11,7 @@ tests:
     platform_allow: nrf52840dk_nrf52840
     tags: nrf52 soc_flash_nrf
     extra_args: OVERLAY_CONFIG=boards/nrf52840_flash_soc.conf
```
this line can be removed CONFIG_MPU_ALLOW_FLASH_WRITE is selected if flash driver is enabled since #20314 was merged.

ikmdani replied:
nrf52840_flash_soc.conf file is enabling CONFIG_MPU_ALLOW_FLASH_WRITE, so you are suggesting only line#13 to  be removed right?

nvlsianpu replied:
I rather think you can merge `drivers.flash.soc_flash_nrf` and `drivers.flash.stm32` testcases to more generic one:

de-nordic commented:
As @nvlsianpu has pointed out, this test should probably be moved to use only storage partition and overlays to move it from internal soc to external flash.
Such change should also enable to figure out the flash device out from DTS in compile time without device specific ifdefs; in similar way the default flash map fills in the structures without knowing the device it is complied for by just figuring out the flash controller as a parent for the partition.

ikmdani commented:
> As @nvlsianpu has pointed out, this test should probably be moved to use only storage partition and overlays to move it from internal soc to external flash.

Do you advise to make this test specific to external flash ?



> Such change should also enable to figure out the flash device out from DTS in compile time without device specific ifdefs; in similar way the default flash map fills in the structures without knowing the device it is complied for by just figuring out the flash controller as a parent for the partition.

There are 2 specific configurations (Nordic QSPI and trusted environment) that I won't be able to test. Considering I'm not yet deep into Zephyr, I'm not comfortable on making too much complex changes to the code base.
Can this be left out for a future change ?

erwango commented:
@nvlsa

> > As @nvlsianpu has pointed out, this test should probably be moved to use only storage partition and overlays to move it from internal soc to external flash.
> 
> Do you advise to make this test specific to external flash ?
> 
> > Such change should also enable to figure out the flash device out from DTS in compile time without device specific ifdefs; in similar way the default flash map fills in the structures without knowing the device it is complied for by just figuring out the flash controller as a parent for the partition.
> 
> There are 2 specific configurations (Nordic QSPI and trusted environment) that I won't be able to test. Considering I'm not yet deep into Zephyr, I'm not comfortable on making too much complex changes to the code base.
> Can this be left out for a future change ?

@nvlsianpu @de-nordic Would you mind provide your views ?
FYI, we'd like to be able to use this test for generic flash testing. Then being soc based of (q)spi-based would depend on board configuration (either default or using overlay).
What is you view this ?

ikmdani committed:
drivers/flash: STM32: A generic change and enabling flash support.

This commit makes the flash test generic. The flast test will
look for storage partition in the soc-flash for testing.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

de-nordic commented on tests/drivers/flash/testcase.yaml:
```diff
@@ -14,3 +14,7 @@ tests:
     tags: mcux
+  drivers.flash.stm32:
+    filter: dt_compat_enabled("st,stm32-flash-controller") and
+       dt_compat_enabled_with_label("fixed-partitions", "storage_partition")
```
Note that all of stm32 devices that have that compatibility, have `flash0` labeled device, instead of this line you may basically, for devices filtered by the compatibility, add universal overlay that would reorganize partitions, or would be adding the required "storage_partition".

de-nordic reviewed (approved):
Note left regarding filtering.

Looks OK to me.
